test/ReverseProxy.Tests/Configuration/ConfigurationConfigProviderTests.cs

```diff
@@ -0,0 +1,572 @@
+            var builder = new ConfigurationBuilder();
+            using var stream = new MemoryStream(Encoding.UTF8.GetBytes(_validJsonConfig));
+            var proxyConfig = builder.AddJsonStream(stream).Build();
+            var services = new ServiceCollection();
+            services.AddSingleton<IProxyConfigProvider, ConfigurationConfigProvider>();
+            var certLoader = new Mock<ICertificateConfigLoader>(MockBehavior.Strict);
+            var certificate = TestResources.GetTestCertificate();
+            certLoader.Setup(l => l.LoadCertificate(It.Is<CertificateConfigData>(o => o.Path == "mycert.pfx" && o.Password == "myPassword1234"))).Returns(certificate);
+            services.AddSingleton(certLoader.Object);
+            services.AddSingleton(new Mock<ILogger<ConfigurationConfigProvider>>().Object);
+            services.Configure<ConfigurationData>(proxyConfig);
+            var serviceProvider = services.BuildServiceProvider();
+
+            var provider = serviceProvider.GetRequiredService<IProxyConfigProvider>();
+            Assert.NotNull(provider);
+            var abstractConfig = provider.GetConfig();
+
+            VerifyValidAbstractConfig(_validConfigurationData, certificate, abstractConfig);
+        }
+
+        [Fact]
+        public void GetConfig_ValidConfiguration_ConvertToAbstractionsSuccessfully()
+        {
+            var certificate = TestResources.GetTestCertificate();
+            var provider = GetProvider(_validConfigurationData, "mycert.pfx", "myPassword1234", () => certificate);
+            var abstractConfig = provider.GetConfig();
+
+            VerifyValidAbstractConfig(_validConfigurationData, certificate, abstractConfig);
+        }
+
+        [Fact]
+        public void GetConfig_ValidConfiguration_AllAbstractionsPropertiesAreSet()
+        {
+            var certificate = TestResources.GetTestCertificate();
+            var abstractionsNamespace = typeof(Abstractions.Cluster).Namespace;
+            var provider = GetProvider(_validConfigurationData, "mycert.pfx", "myPassword1234", () => certificate);
+            var abstractConfig = (ConfigurationSnapshot) provider.GetConfig();
+            //Removed incompletely filled out instances.
+            abstractConfig.Clusters = abstractConfig.Clusters.Where(c => c.Id == "cluster1").ToList();
+            abstractConfig.Routes = abstractConfig.Routes.Where(r => r.RouteId == "routeA").ToList();
+
+            VerifyAllPropertiesAreSet(abstractConfig);
+
+            void VerifyFullyInitialized(object obj, string name)
+            {
+                switch (obj)
+                {
+                    case null:
+                        Assert.True(false, $"Property {name} is not initialized.");
+                        break;
+                    case Enum m:
+                        Assert.NotEqual(0, (int)(object)m);
+                        break;
+                    case string str:
+                        Assert.NotEmpty(str);
```

Comment:
Isn't NotEmpty for lists? Or does this work because you already checked for null above.
```suggestion
                        Assert.False(String.IsNullOrEmpty(str));
```

Reply:
`string` [implements IEnumerable](https://docs.microsoft.com/en-us/dotnet/api/system.string?view=netcore-3.1), thus `Assert.Empty()` works fine.
```C#
Assert.Empty("");
Assert.Empty(string.Empty);
```
I chose this form because it's shorter than `Assert.False(String.IsNullOrEmpty(str))`